src/notification/index.js

```diff
@@ -247,17 +234,22 @@ export class NotificationListener {
       this.listenAndroid('notificationOpened', this.handleNotificationOpen);
       this.listenAndroid('remoteNotificationsRegistered', this.handleDeviceToken);
     } else {
-      // On iOS, `note` should be an IOSNotifications object. The notification
-      // data it returns from `getData` is unvalidated -- it comes almost
-      // straight off the wire from the server.
-      this.listenIOS('notificationOpened', (note: { getData(): JSONableDict }) => {
-        const data = fromAPNs(note.getData());
-        if (data) {
-          this.handleNotificationOpen(data);
+      this.listenIOS('notification', (notification: PushNotificationIOS) => {
+        // This is actually typed as ?Object (and so effectively `any`); but if
+        // present, it must be a JSONable dictionary. (See PushNotificationIOS.js and
+        // RCTPushNotificationManager.m in Libraries/PushNotificationIOS.)
+        const data: ?JSONableDict = notification.getData();
+        if (!data) {
+          return;
         }
+        const dataFromAPNs: Notification | void = fromAPNs(data);
```

Comment:
> (and returning `undefined` when invalid)

It also throws a special kind of error sometimes; it says `@throws An ApnsMsgValidationError on interesting failure.`. Would you say a nullish `data` is an interesting failure? I'm not so sure what would be going on to make `data` nullish.

Reply:
Yeah, more precisely: it looks like `fromAPNsImpl` has that `@throws`, and `fromAPNs` catches such errors, logs them, and returns `undefined`.

I'm not sure when that would happen either. That means it'd be sensible to log a warning in that case, just like we do for malformed data blobs. Probably "interesting" wasn't a very clear way to write that description; the distinction that I think it's intended to get at is "unexpected", as in "buggy".

That is:
```
// @returns A `Notification` on success, `undefined` on suppressible failure.
// @throws An ApnsMsgValidationError on interesting failure.
```
* it returns `undefined` when the given data has a form that we know happens, and we understand why, but the app can't currently do anything with.

  There's one such case now, which is when the notification-message is of type `remove` or some unknown future type, rather than type `message` (or no type, which implicitly means `message`.)
* it throws an error when the given data has a form that we don't expect -- meaning there's a bug either in the app or in the server.

And then its caller logs a warning in the latter case -- indicating a bug -- and in both cases returns `undefined` so we do nothing further with the notification.